### PR TITLE
Iam condition samples

### PR DESCRIPTION
**Pull request by frankyn**
I'm using a specific version of java-storage library. Need to remove it.

### CODE REVIEW

File: google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/ListBucketIamMembers.java [hunk 1]
```diff
@@ -34,10 +32,15 @@ public static void listBucketIamMembers(String projectId, String bucketName) {
...
     Storage storage = StorageOptions.newBuilder().setProjectId(projectId).build().getService();
     Policy policy = storage.getIamPolicy(bucketName);
-    Map<Role, Set<Identity>> policyBindings = policy.getBindings();
+    List<Binding> policyBindings = policy.getBindingsList();
 
-    for (Map.Entry<Role, Set<Identity>> entry : policyBindings.entrySet()) {
-      System.out.printf("Role: %s Identities: %s\n", entry.getKey(), entry.getValue());
+    for (Binding binding : policy.getBindingsList()) {
+      System.out.printf("Role: %s Identities: %s\n", binding.getRole(), binding.getMembers());
+      if (null != binding.getCondition()) {
```
Comment (kolea2) [inline google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/ListBucketIamMembers.java]: swap

File: google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/RemoveBucketIamConditionalBinding.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        String role = "roles/storage.objectViewer";
+
+        List<Binding> bindings = new ArrayList(originalPolicy.getBindingsList());
+        Condition.Builder conditionBuilder = Condition.newBuilder();
+        conditionBuilder.setTitle("Title");
+        conditionBuilder.setDescription("Description");
+        conditionBuilder.setExpression(
+                "resource.name.startsWith(\"projects/_/buckets/bucket-name/objects/prefix-a-\")");
+
+        Iterator iterator = bindings.iterator();
```
Comment (kolea2) [inline google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/RemoveBucketIamConditionalBinding.java]: what is this doing? I'd add a comment

File: google-cloud-examples/pom.xml [hunk 3]
```diff
@@ -76,6 +76,7 @@
     <dependency>
       <groupId>com.google.cloud</groupId>
       <artifactId>google-cloud-storage</artifactId>
+      <version>1.105.0</version>
```
Comment (kolea2) [inline google-cloud-examples/pom.xml]: when the libraries-bom gets updated, we can remove this version number

File: google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/RemoveBucketIamMember.java [hunk 4]
```diff
@@ -35,21 +35,34 @@ public static void removeBucketIamMember(String projectId, String bucketName) {
...
+    boolean bindingExists = false;
+    for (Binding binding : updatedPolicy.getBindingsList()) {
+      if (binding.getRole().equals(role) && null == binding.getCondition()) {
+        bindingExists = true;
+        if (!binding.getMembers().contains(member)) {
+          System.out.printf("Removed %s with role %s from %s\n", member, role, bucketName);
+        }
+      }
+    }
+    if (!bindingExists) {
```
Comment (kolea2) [inline google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/RemoveBucketIamMember.java]: do we want to have an else here? Are you expected to remove a role? Should it throw?

File: google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/RemoveBucketIamMember.java [hunk 5]
```diff
@@ -35,21 +35,34 @@ public static void removeBucketIamMember(String projectId, String bucketName) {
...
+    String role = "roles/storage.objectViewer";
 
-    // See the Identity documentation for other identities:
-    // https://googleapis.dev/java/google-cloud-core/latest/com/google/cloud/Identity.html
-    Identity identity = Identity.group("example@google.com");
+    String member = "group:example@google.com";
+    List<Binding> bindings = new ArrayList(originalPolicy.getBindingsList());
+    for (int index = 0; index < bindings.size(); ++index) {
+      Binding binding = bindings.get(index);
+      if (binding.getRole().equals(role)
```
Comment (kolea2) [inline google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/RemoveBucketIamMember.java]: I'd extract this as a descriptive variable boolean

**Comment by kolea2** [comment]
Added some comments, some apply in more than one area if the code is duplicated elsewhere in another example

File: google-cloud-examples/pom.xml (hunk 3 above)
Comment (frankyn) [reply google-cloud-examples/pom.xml]: Is it okay to move forward with it for now?

File: google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/AddBucketIamConditionalBinding.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+    Storage storage = StorageOptions.newBuilder().setProjectId(projectId).build().getService();
+    Policy originalPolicy =
+        storage.getIamPolicy(bucketName, Storage.BucketSourceOption.requestedPolicyVersion(3));
+
+    String role = "roles/storage.objectViewer";
+
+    String member = "group:example@google.com";
+
+    List<Binding> bindings = new ArrayList(originalPolicy.getBindingsList());
```
[the remark this answers is not in the document]
Comment (frankyn) [reply google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/AddBucketIamConditionalBinding.java]: Return type is immutablelist.

File: google-cloud-examples/pom.xml [hunk 7]
```diff
@@ -76,6 +76,12 @@
     <dependency>
       <groupId>com.google.cloud</groupId>
       <artifactId>google-cloud-storage</artifactId>
+      <version>1.104.1-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (frankyn) [reply google-cloud-examples/pom.xml]: libraries-bom needs to be updated.

File: google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/AddBucketIamConditionalBinding.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (jkwlui) [reply google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/AddBucketIamConditionalBinding.java]: Let's make that clear by replacing the comment with:

`// getBindingsList() returns an ImmutableList. Copying over as an ArrayList so it's modifiable.`

File: google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/AddBucketIamConditionalBinding.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+            .setRole(role)
+            .setMembers(Arrays.asList(member))
+            .setCondition(conditionBuilder.build())
+            .build());
+
+    // Update policy with new conditional binding
+    Policy updatedPolicy =
+        storage.setIamPolicy(
+            bucketName,
+            originalPolicy.toBuilder().setBindings(bindings).setVersion(policyVersion).build());
```
Comment (jkwlui) [inline google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/AddBucketIamConditionalBinding.java]: (style): I would pull this line out to a variable before the call so it highlights the need to re-build the policy:

```java
Policy.Builder builder = originalPolicy.toBuilder();
Policy policyToSet = builder
    .setBindings(bindings)
    .setVersion(policyVersion)
    .build();

Policy updatedPolicy = storage.setIamPolicy(bucketName, policyToSet);
```

File: google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/AddBucketIamConditionalBinding.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+    System.out.printf(
+            "Added conditional binding with role %s to %s\n", member, role, bucketName);
```
Comment (jkwlui) [inline google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/AddBucketIamConditionalBinding.java]: Following the canonical samples, we should also print out the condition title, description and expression :)
```ruby
puts "Added #{member} with role #{role} to #{bucket_name} with condition #{title} #{description} #{expression}"
```

File: google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/AddBucketIamMember.java [hunk 10]
```diff
@@ -32,24 +33,33 @@ public static void addBucketIamMember(String projectId, String bucketName) {
...
 
+    // For more information please read:
+    // https://cloud.google.com/storage/docs/access-control/iam
     Storage storage = StorageOptions.newBuilder().setProjectId(projectId).build().getService();
-    Policy originalPolicy = storage.getIamPolicy(bucketName);
 
-    // See the StorageRoles documentation for other valid roles:
-    // https://googleapis.dev/java/google-cloud-clients/latest/com/google/cloud/storage/StorageRoles.html
-    Role role = StorageRoles.objectViewer();
+    int policyVersion = 3;
```
Comment (jkwlui) [inline google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/AddBucketIamMember.java]: It got me thinking here - we want developers to use version `3` everywhere. Would it be better from that standpoint to get rid of this variable and directly use the value everywhere?

```java
...BucketSourceOption.requestedPolicyVersion(3);
```

```java
builder.setVersion(3);
```

Comment (frankyn) [reply google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/AddBucketIamMember.java]: It was recommended by @kolea2 to use a separate variable. I'm okay with either but it's easier to understand what the value 3 means.

File: google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/RemoveBucketIamMember.java [hunk 11]
```diff
@@ -31,26 +31,41 @@ public static void removeBucketIamMember(String projectId, String bucketName) {
...
+    String role = "roles/storage.objectViewer";
+    String member = "group:example@google.com";
 
-    // See the Identity documentation for other identities:
-    // https://googleapis.dev/java/google-cloud-core/latest/com/google/cloud/Identity.html
-    Identity identity = Identity.group("example@google.com");
+    // Get policy bindings list as a mutable ArrayList.
+    List<Binding> bindings = new ArrayList(originalPolicy.getBindingsList());
 
+    // Remove member from binding
```
Comment (jkwlui) [inline google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/RemoveBucketIamMember.java]: `// Remove role-member binding without a condition.`

File: google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/AddBucketIamConditionalBinding.java (hunk 8 above)
Comment (frankyn) [reply google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/AddBucketIamConditionalBinding.java]: good idea!

File: google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/AddBucketIamMember.java (hunk 10 above)
Comment (kolea2) [reply google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/AddBucketIamMember.java]: I'll defer to you both here - please feel free to use whichever you think would make more sense to the user :)

File: google-cloud-examples/pom.xml (hunk 3 above)
Comment (kolea2) [reply google-cloud-examples/pom.xml]: yep fine with me!

File: google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/RemoveBucketIamMember.java [hunk 12]
```diff
@@ -31,26 +31,40 @@ public static void removeBucketIamMember(String projectId, String bucketName) {
...
 
-    Policy updatedPolicy =
-        storage.setIamPolicy(
-            bucketName, originalPolicy.toBuilder().removeIdentity(role, identity).build());
+    // Remove role-member binding without a condition.
+    for (int index = 0; index < bindings.size(); index++) {
+      Binding binding = bindings.get(index);
+      boolean foundRole = binding.getRole().equals(role);
+      boolean foundMember = binding.getMembers().contains(member);
+      boolean bindingIsNotConditional = (binding.getCondition() == null);
```
Comment (kolea2) [inline google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/RemoveBucketIamMember.java]: nit - remove `()` around the boolean

File: google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/AddBucketIamMember.java (hunk 10 above)
Comment (frankyn) [reply google-cloud-examples/src/main/java/com/google/cloud/examples/storage/buckets/AddBucketIamMember.java]: Following canonical samples on this one which aligns with @jkwlui.

**Pull request merged**